Add "Pebble Notices" section to Pebble doc

benhoyt opened the pull request:
This PR adds a section about "Pebble Notices" to [this doc](https://juju.is/docs/sdk/interact-with-pebble), with charm and test examples.

It also adds a very basic `pebble-custom-notice` event page, based on the [pebble-ready one](https://juju.is/docs/sdk/container-name-pebble-ready-event).

tonyandrewmeyer commented on sdk/interact-with-pebble.md:
```diff
@@ -831,6 +832,96 @@ Traceback (most recent call last):
+
+<a href="#heading--use-custom-notices-from-the-workload-container"><h2 id="heading--use-custom-notices-from-the-workload-container">Use custom notices from the workload container</h2></a>
+
+From version 3.4, Juju includes support for [Pebble Notices](https://github.com/canonical/pebble/#notices). Juju polls each workload container's Pebble server for new notices, and fires an event to the charm when a notice first occurs as well as each time it repeats.
```
Maybe "first occurs as well as when it repeats"? I assume you don't want to talk about squashing repeated notices here, but saying "each time" seems to imply that it is every time?

tonyandrewmeyer replied:
Hmm, have read further down now. I guess the squashing is preventing it *repeating* and so "each time it repeats" is actually correct. So perhaps ignore this, but leaving it here in case my thought process as I read through it is useful.

tonyandrewmeyer commented on sdk/interact-with-pebble.md:
```diff
@@ -831,6 +832,96 @@ Traceback (most recent call last):
+
+From version 3.4, Juju includes support for [Pebble Notices](https://github.com/canonical/pebble/#notices). Juju polls each workload container's Pebble server for new notices, and fires an event to the charm when a notice first occurs as well as each time it repeats.
+
+Each notice has a *type* and *key*, the combination of which uniquely identifies it. A notice's count of occurrences is incremented every time a notice with that type and key combination occurs.
```
This doesn't need to be in the doc, but for my edification: does this reset? Is it per-container and lives as long as the container lives? Or it's global (to the model?) and lives as long as the model lives?

benhoyt replied:
No, this doesn't reset. It's persisted to disk on the container, but realistically a container is kind of ephemeral, so it will only exist as long as the container lives.

tonyandrewmeyer commented on sdk/interact-with-pebble.md:
```diff
@@ -831,6 +832,96 @@ Traceback (most recent call last):
+
+Currently, the only supported notice type is "custom". These are custom notices recorded by a user of Pebble; in future, other notice types may be recorded by Pebble itself. When a custom notice occurs, Juju fires a [`PebbleCustomNoticeEvent`](https://ops.readthedocs.io/en/latest/#ops.PebbleCustomNoticeEvent) event whose [`workload`](https://ops.readthedocs.io/en/latest/#ops.WorkloadEvent.workloa ...
+
+Custom notices allow the workload to wake up the charm when something interesting happens on the workload, for example, when a PostgreSQL backup process finishes, or some kind of alert occurs.
```
Nit: "on the workload" sounds odd to my ears. "on the container" or "on the workload container" or "with the workload"?

benhoyt replied:
Updated to "with the workload", thanks.

tonyandrewmeyer commented on sdk/interact-with-pebble.md:
````diff
@@ -831,6 +832,96 @@ Traceback (most recent call last):
+
+```python
+class PostgresCharm(ops.CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
````
Unrelated to notices, but we'd talked about moving towards this instead:

```suggestion
    def __init__(self, framework: ops.Framework):
        super().__init__(framework)
```

benhoyt replied:
Good idea, updated. And also created https://github.com/canonical/charmcraft/pull/1477 to update the `charmcraft init` templates.

tonyandrewmeyer reviewed (commented):
This looks great to me - only one very tiny nit.

tmihoc reviewed (changes requested):
I have some suggestions. They'are essentially all about deep Diataxis -- making sure that:
- every new concept / tool has its proper reference and is only linked to from the how-to where needed
- the how-to, which is actually a collection of how-tos, has a clear structure that reflects that

benhoyt commented:
To move things forward, I'm going to merge this as is (I've updated the [main doc](https://discourse.charmhub.io/t/how-to-run-workloads-with-a-charm-kubernetes/4554) in Discourse and added the [event one](https://discourse.charmhub.io/t/event-container-pebble-custom-notice/13019)), and follow up with the changes Dora suggested in a meeting with her so we can do that together.
